### PR TITLE
[threadpool-io] Move IOSelector to manage

### PR DESCRIPTION
**Pull request by ludovic-henry**

### CODE REVIEW

**Comment by ludovic-henry** [comment]
build

**Comment by ludovic-henry** [comment]
build

File: mono/utils/mono-threads-posix-signals.c [hunk 1]
```diff
@@ -44,6 +44,8 @@ signal_search_alternative (int min_signal)
 	/* we try to avoid SIGRTMIN and any one that might have been set already, see bug #75387 */
 	for (i = MAX (min_signal, SIGRTMIN) + 1; i < SIGRTMAX; ++i) {
 		struct sigaction sinfo;
+		if (i == DEFAULT_SUSPEND_SIGNAL || i == DEFAULT_RESTART_SIGNAL)
```
Comment (kumpera) [inline mono/utils/mono-threads-posix-signals.c]: This is weird, looks unrelated.

**Comment by ludovic-henry** [comment]
build

File: mcs/class/System/System/IOSelector.cs [hunk 2]
```diff
@@ -155,10 +157,376 @@ public void MarkDisposed ()
...
+							}
+							case UpdateType.Add: {
+								bool exists = states.ContainsKey (u.handle);
+								if (!exists)
+									states.Add (u.handle, new List<IOSelectorJob> ());
+
+								states [u.handle].Add (u.job);
+
+								int idx;
+								if ((idx = backend_updates.FindLastIndex (bu => bu.handle == u.handle)) == -1) {
```
Comment (kumpera) [inline mcs/class/System/System/IOSelector.cs]: This introduces a heap allocated closure.

File: mcs/class/System/System/IOSelector.cs [hunk 3]
```diff
@@ -155,10 +157,376 @@ public void MarkDisposed ()
...
+										case BackendUpdateType.Add:
+											backend_updates.RemoveAt (idx);
+											if (!bu.is_new)
+												backend_updates.Add (new BackendUpdate (u.handle, BackendUpdateType.Remove));
+
+											break;
+										case BackendUpdateType.Remove:
+											throw new InvalidOperationException (String.Format ("Cannot remove handle {0} twice", (int) u.handle));
+										default:
+											throw new InvalidOperationException ();
```
Comment (kumpera) [inline mcs/class/System/System/IOSelector.cs]: You probably want to add some text here on what was the state.

File: mcs/class/System/System/IOSelector.cs [hunk 4]
```diff
@@ -155,10 +157,376 @@ public void MarkDisposed ()
...
+										}
+									}
+								}
+
+								for (int j = i + 1; j < updates_last; ++j) {
+									if (updates [j].type == UpdateType.Add && updates [j].handle == u.handle) {
```
Comment (kumpera) [inline mcs/class/System/System/IOSelector.cs]: This is quadratic on the number of updates. Is this an issue?

File: mcs/class/System/System/IOSelector.cs [hunk 5]
```diff
@@ -155,10 +157,376 @@ public void MarkDisposed ()
...
+
+						backend_updates.Clear ();
+
+						Monitor.PulseAll (updates);
+					}
+
+					if (backend_events.Length < states.Count)
+						Array.Resize (ref backend_events, states.Count);
+
+					for (int i = 0; i < backend_events.Length; ++i)
```
Comment (kumpera) [inline mcs/class/System/System/IOSelector.cs]: This looks wrong and wastefull.

Why can't BackendPoll return the number of events it found?

This way you avoid zero'ing data before and looping over the whole thing later.

File: mcs/class/System/System/IOSelector.cs [hunk 6]
```diff
@@ -155,10 +157,376 @@ public void MarkDisposed ()
...
+							}
+						}
+
+						backend_updates.Clear ();
+
+						Monitor.PulseAll (updates);
+					}
+
+					if (backend_events.Length < states.Count)
+						Array.Resize (ref backend_events, states.Count);
```
Comment (kumpera) [inline mcs/class/System/System/IOSelector.cs]: This looks wasteful. We won't have updates for all sockets on every polling cycle.

We should dynamically grown every time BackendPoll find more events that slots to return them.

File: mcs/class/System/System/IOSelector.cs (hunk 2 above)
Comment (kumpera) [inline mcs/class/System/System/IOSelector.cs]: And it's quadratic on the number of updates, not great either.

**Comment by kumpera** [comment]
The IO threadpool code have very little test coverage and this is just adding more instability to the existing code. This change is a rewrite from what we have today.

We should first get better tests for that part of the codebase, exercising the many entry points before we more into merging it.

The current code looks very fragile to implementation and user mistakes, possibly taking down the whole IOTP in the process.

We should get such a beast up and running first before getting into rewriting more code.

**Comment by lewurm** [comment]
is there still interest to merge this or should it be closed?
